### DISCUSSION TITLE
Async command parsing in ImageSharp.Web

### DISCUSSION THREAD

**Discussion by wagich** [SixLabors/ImageSharp.Web#99] [Ideas] [answered]
I have a use case where I need to set a default focal/anchor point (parameter `rxy`) when processing an image. This focal point data is stored in a database and it would be beneficial to have command parsing be async to allow for async db access.

Since most calls would complete synchronously, the API is returning ValueTask to avoid allocating a task when it's not needed.
```c#
public interface SixLabors.ImageSharp.Web.Commands.IRequestParser
{
    // existing API
    // IDictionary<string, string> ParseRequestCommands(HttpContext context);

    // new API
    ValueTask<IDictionary<string, string>> ParseRequestCommandsAsync(HttpContext context);
}
```
I've implemented it in https://github.com/wagich/ImageSharp.Web/tree/async-command-parsing
Would you be open to a PR for this?

**Comment by JimBobSquarePants** [comment, marked as answer]
Hi @wagich that sounds like something you would do via the [OnParseCommands](https://docs.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddlewareOptions.html#SixLabors_ImageSharp_Web_Middleware_ImageSharpMiddlewareOptions_OnParseCommands) event.

Do you have a single focal point reused across all images or one per image?

**Reply by wagich** [reply]
Hmm, I had looked at `OnParseCommands` but decided I couldn't use it and needed to use a custom async `IRequestParser`. But right now I can't reconstruct why… Like you said, it doesn't seem to be neccessary.

I have a focal point per image. Here's the code that runs to retrieve it from the db and set it as a command:
```c#
// add focal point from asset metadata unless one is specified in the url
if (!commands.ContainsKey(ResizeWebProcessor.Xy))
{
	// we should be endpoint-routed, get id from route data
	if (Id.TryParse(context.GetRouteValue("id") as string, out var assetId))
	{
		// since we're in a singleton, we need to service-locate the db session from HttpContext
		var session = context.RequestServices.GetRequiredService<IAsyncDocumentSession>();
		var asset = await session.LoadAsync<AssetBase>(assetId).ConfigureAwait(false);
		if (asset is IImageAsset imageAsset)
		{
			var x = imageAsset.Crop.FocalPoint.X;
			var y = imageAsset.Crop.FocalPoint.Y;
			commands.Add(ResizeWebProcessor.Xy, FormattableString.Invariant($"{x},{y}"));
		}
	}
	else
	{
		throw new InvalidOperationException("Route data value 'id' is null or not an Id. Make sure the ImageSharp middleware is registered inside the correct endpoint.");
	}
}
```

This indicates that just an async version of `OnParseCommands` would be fine. Im _very_ wary of doing sync-over-async however…

**Reply by JimBobSquarePants** [reply]
I'm curious? Why do you not pass the default value via the URL?

**Reply by wagich** [reply]
That was my first impulse. The short answer is that I want to avoid the db roundtrip(s) when rendering the output and this seemed like a good alternative.

The images are referenced in formatted text, and there could be maybe 20-40 images in a single output. The references contain Id, Name, Description etc., everything needed to construct the image URL. Resolution and crop-mode are passed in from the render function.

Adding the focal point at render-time would involve hitting the db once for each image (not good) or pre-parsing the text and then batch-loading metadata for all images in a single trip (complicating the code).

In the context of the subsequent image file requests however, the metadata is already loaded (since I need to check if it exists, is an image etc.), which means I don't need to make any additional db roundtrips at all… It also speeds up the initial render by deferring the work to the image request itself.

**Reply by wagich** [reply]
And maybe to give a bit more context:
I'm using a custom `IImageProvider`/`IImageResolver` that does the metadata-loading and validation. It also handles cache-invalidation when the metadata changes. Adding commands like this would lead to problems with caching otherwise, I'd think.

**Reply by JimBobSquarePants** [reply]
Ok... That all makes sense. So it seems like converting the `Action OnParseCommands` to `Func<Task> OnParseCommandsAsync` and do the same for other events might be the best approach?

**Reply by wagich** [reply]
It looks like it. I'll take some time today and change my code to use the `On…` just to verify that it works.

**Reply by JimBobSquarePants** [reply]
Excellent. Thanks!

**Reply by wagich** [reply]
It seems to be working using the callback. I've opened a draft pull request #100 with the changes.
Let me know if I need to change anything or please feel free to close the PR if you prefer a different approach.
